willow-eventhandler/src/main/java/com/nitorcreations/willow/eventhandler/EventPoller.java

```diff
@@ -0,0 +1,231 @@
+          // Clear the previous message IDs and set the now received message IDs, note the synchronization
+          messageIds.clear();
+          messageIds.addAll(newMessageIds);
+        }
+        // Loop through configured handlers to handle the non-duplicate message
+        for (EventMessage eventMessage : uniqueEventMessages) {
+          List<EventHandler> messageEventHandlers = eventHandlers.get(eventMessage.eventType);
+          if (messageEventHandlers != null) {
+            for (EventHandler eventHandler : messageEventHandlers) {
+              try {
+                eventHandler.handle(eventMessage);
+              } catch (Exception e) {
+                logger.log(Level.INFO, "Failure in handling event data", e);
+              }
+            }
+          }
+        }
+      }
+    }
+
+    @OnWebSocketClose
+    public void handleClose(Session closedSession, int statusCode, String reason) {
+      logger.log(Level.INFO, "Event listener websocket closed with status " + statusCode + ". Reason: " + reason);
+      closeClient();
+      connectLatch.countDown();
+    }
+
+    @OnWebSocketError
+    public void handleError(Session errorSession, Throwable throwable) {
+      logger.log(Level.INFO, "Event listener websocket error", throwable);
+      closeClient();
+      connectLatch.countDown();
+    }
+
+    @OnWebSocketConnect
+    public void startPolling(Session session) {
+      logger.info("Sending poll request for events");
+      MetricConfig metricConfig = new MetricConfig();
+      long now = System.currentTimeMillis();
+      metricConfig.setId("eventhandler");
+      metricConfig.setMetricKey("/event");
+      metricConfig.setStart(now - 20000);   // For the first chunk, send events from 20s ago
+      metricConfig.setStop(now);            // ... up to the current moment.
+      metricConfig.setStep(10000);          // Then keep sending data every 10 seconds.
+      metricConfig.setMinSteps(2);          // Every time send data for the last 2 * 10 = 20 seconds.
+      String json = gson.toJson(metricConfig);
+      logger.info("Sending polling request " + json);
+      try {
+        session.getRemote().sendString(json);
+      } catch (IOException e) {
+        logger.log(Level.SEVERE, "Failed to send event polling request", e);
+        closeClient();
+        connectLatch.countDown();
+      }
+    }
+
+    @Override
+    public void run() {
+      while (!Thread.currentThread().isInterrupted()) {
+        WebSocketClient client = getClient();
+        if (!client.isRunning() && !client.isStarting() || client.isFailed()) {
+          try {
+            connect();
+            connectLatch = new CountDownLatch(1);
+          } catch (Exception e) {
+            logger.log(Level.SEVERE, "Error setting up websocket connection", e);
+          }
+        }
+        try {
+          connectLatch.await();
+        } catch (InterruptedException e) {
+        }
+      }
+    }
+
+    private void closeClient() {
```

Comment:
every caller to closeClient also does connectLatch.countDown() maybe move the countDown into closeClient?